### ISSUE TITLE
How can i still receive messages when the Signal application is closed?

### ISSUE THREAD

**Issue by kepler2** [WhisperSystems/Signal-Android#6096]
I have:
- [x] searched open and closed issues for duplicates
- [x] read https://github.com/WhisperSystems/Signal-Android/wiki/Submitting-useful-bug-reports

----------------------------------------

 Description:

I cannot really call it a bug, more of a question. 

 **Steps to reproduce**
- Register a Signal account > Complete the registration
- Register Signal on another device and complete the registration
- Close the Signal application (with password enabled - but occurs also without password enabled) 
- Send messages to the device with the Signal application closed

**Actual result**: Well, like in the title, from what I've saw is that the messages are not received until the Signal application is opened. Is this the intended behavior? I mean I also have WhatsApp installed that can receive messages even when the application is closed from the task switcher. (it seems that it has a background process / service working, maybe this is why?)

Note that i have Notifications enabled, Background Data is not restricted, I have good Internet connection and using Android 5.1.1.

**Expected result**: Signal app should receive notifications / messages even when the application is closed.

Thanks

Device info:

**Device:** Asus Zenfone Go
**Android version:** 5.1.1
**Signal version:** 3.27.1

**Comment by Trolldemorted** [comment]
Signal should receive messages even when the app is "not open".

Does it *always* fail to notice incoming messages? Did you try rebooting your device?

**Comment by FeuRenard** [comment]
You can try this:
https://support.whispersystems.org/hc/en-us/articles/213190487-Why-is-there-a-delay-in-receiving-messages-

**Comment by kepler2** [comment]
@Trolldemorted I can reproduce easily by performing the following steps:

1. Open phone A, enable 3g connectivity launch Signal.
2. Open phone B, enable 3g connectivity, launch Signal.
3. Send a message from phone A to phone B - observe that the messages are received without issues.
4. Disable connectivity on phone B and close the application using the task switcher. (Also lock it in case you have password)
5. Send a message from phone A to phone B.
6. Enable connectivity on phone B.
7. Do not open Signal on device B.

Observe that the messages are received on device B only after the Signal application is opened.

So basically, I need to open the app whenever I enable connectivity or else the messages are not received.

**Comment by Trolldemorted** [comment]
Thanks, can indeed reproduce it (Android6, redmi note 3).

Sometimes (read: rarely) a new incoming message (or syncmessage, or syncreadmessage?) appears to kickstart the receiving. Maybe ill test it more thoroughly when there is less noise on my signal account.

Since whatsapp can cope with this, could it be a signal-server bug?

**Comment by 2-4601** [comment]
There's an issue somewhere about closing the app from the overview screen and then not receiving notifications afterwards. (Don't have energy to find it now.)

**Comment by 2-4601** [comment]
Closing as duplicate of #4965
(@kepler2 could you add your reproduction steps to that issue?)

**Comment by Trolldemorted** [comment]
@2-4601 why should this be a duplicate? This issue is about signal not noticing the establishment of a network connection, not about signal being killed on swipe.

My signal works fine after swiping if the connection is stable. Please reopen.

**Comment by 2-4601** [comment]
@Trolldemorted Because the original reproduction steps and the steps added later list the closing from the recents view. You also replied that you can reproduce it and did not say until now that it is also reproducible without explicit closing.

**Comment by Trolldemorted** [comment]
It is not reproducible without explicit "closing", but, as i said, this is not #4956 since as long as the connection is stable messages are received just fine. #4956 was classified as a ROM issue, whereas this looks like a Signal/GMS/Signal-Server issue to me, or is this not reproducible on your phone?

**Comment by kepler2** [comment]
Can I open another issue with the updated reproduction steps? In my opinion this is dealbreaking as it affects the usability of the app.

For me at least I need to instruct friends to keep the app opened all times, or to open it at least once in order to receive messages.

I really appreciate all the hard work that is put into this app. Thanks for the replies and sry if I don't appear tech savvy but I  try to explain as good as I can.

**Comment by 2-4601** [comment]
Cannot reproduce on Moto G4 Play, Android 6.0.1, Signal 3.27.1

1. I disable Wi-Fi and enable cellular data
1. While Signal is in the background someone sends me a Signal message and I receive a notification of it as expected.
1. I disable cellular data
1. I swipe-close Signal from recents
1. Someone sends me a Signal message
1. I enable cellular data
1. After a minute or so I receive a notification for the message

Then I did the same again without step 4 and result is the same.

https://gist.github.com/anonymous/9284a18b6e22f22918f08d8542de943e

I don't think the cause of #4965 is clear. ROM issue was just speculated.

As referenced in #4965 the swiping from recents has had issues with receiving notifications https://github.com/WhisperSystems/Signal-Android/issues/1037#issuecomment-36767446:
> This is the consequence of a workaround for an Android OS bug that was causing TextSecure to miss incoming messages after swiping from the recent tasks list. I agree it's unfortunate, but it's better than the alternative, and I don't think there's anything we can do until Android fixes their bug.

But the state of that OS bug seems to be unclear at the moment as is the fix/workaround that Signal is using.

***

@kepler2 
> For me at least I need to instruct friends to keep the app opened all times, or to open it at least once in order to receive messages.

Do notifications work for you if you don't swipe Signal from recents? If they do, is there some reason you would want to swipe it off?

**Comment by Trolldemorted** [comment]
on a second try: Sometimes i get the notification properly after ~1min, sometimes it just takes several minutes, and i was not that patient during my first attempt.

@kepler2 are you really sure that if you go through 2-4601's steps this problem arises *every* time, and that these messages do not come a few minutes delayed?

Do you have Signal-Desktop instances linked to your Android?

**Comment by kepler2** [comment]
I'll try to test again tomorrow and report here. Thanks.

**Comment by kepler2** [comment]
@2-4601 Well if i restart the smartphone, signal is not present in the recent apps, so I have to manually start it. Today the same, someone sent me a message I did not receive it until i opened the application. I'll test again later.

**Comment by 2-4601** [comment]
@kepler2 Have you checked the [support article](https://support.whispersystems.org/hc/en-us/articles/213190487-Why-is-there-a-delay-in-receiving-messages-) which @FeuRenard linked to?

After a quick search it seems like ZEN UI has some power saving modes that could affect this: https://www.asus.com/zentalk/thread-3135-1-1.html

**Comment by kepler2** [comment]
@2-4601: Thanks you for the guidance. I think I'm on the good lane now but i still have to check on another device. For the Asus I think it has to do with the Auto-Start manager, I think its part of the "bloat" asus ships with its smartphones and i think it puts the newly installed apps on "deny", which means that Signal has some process that starts once the smartphone is booted?

As you can see in the screenshot below, I have put Signal to Allow and now it seems i can receive messages after i enable the internet connection (WIFI works faster, with 3G messages are received with a delay but i guess its not a big problem)

I still have to check on another device to see if the problem really is on my Asus and I'll reply here.

Screenshot:

![screenshot](https://cloud.githubusercontent.com/assets/18610924/22396144/8f8728d4-e55a-11e6-8c23-70ef9335ce13.jpg)

**Comment by kepler2** [comment]
@2-4601 Hello. I've managed to reproduce the issue on the other device which has CyanogenMod 12.1. Note that I've looked for a similar setting as in the Asus device but this has none related. Sry for talking your time...

**Comment by 2-4601** [comment]
Regarding CM 12.1 it could be a bad Google Apps install. The [support article](https://support.whispersystems.org/hc/en-us/articles/213190487-Why-is-there-a-delay-in-receiving-messages-) also mentions that and has a link to Cyanogen wiki which you might want to check out.

And if none of the other options in the support article work we should move on to issue #4965 of which this issue seems to be a duplicate after all.

**Comment by Trolldemorted** [comment]
@kepler2 does this happen *every* time on the cm12 device?

Could you run `ps | grep thought` as root in a terminal/adb shell to see whether the signal process is still alive?

I am still not convinced that this is 4965, since this issue is also occuring after rebooting, and not just after swiping, and only after swiping w/o network.

**Comment by kepler2** [comment]
Last time i checked, i received notifications but with several minutes delay. But when i posted about CM12 i did not receive (or maybe i was too fast in closing the internet connection?), I still have to check thoroughly. 

@Trolldemorted Can I run that command in Linux and also, the process should be visible even after a fresh start of the phone?

**Comment by Trolldemorted** [comment]
You may run the command either on the phone directly in CM's terminal app, or via an adb shell on your pc - adb should run on most linux distributions.

**Comment by kepler2** [comment]
Guys, it seems to work properly, I don't really know that the issue was. Thanks for helping and thanks for this app.
